File: python/miscellaneous_practice/LeetCode_53_Maximum_subarray.py
# prblem statement
# We need to find the subarray that has the maximum sum.
# A subarray is simply an array inside another array.
# ex:-
# [-2, 1, -3, 4, -1, 2, 1, -5, 4]
# Possible subarrays: [-2, 1, -3], [-5, 4], or even the entire array itself.
# key point is that the subarray must be continuous.


# brute force method
# Using two loops to explore every possible subarray.
# We'll use a variable to track the largest sum found so far.
# Note: We only need to return the maximum sum, not the subarray itself.
# T.C -> O(n^2)
# S.C -> O(1)

def brute_force(nums):
    length = len(nums)
    maximum_sum = nums[0]

    for i in range(length):
        current_sum = 0
        for j in range(i, length):
            current_sum += nums[j]
            maximum_sum = max(current_sum, maximum_sum)
    return maximum_sum

# kadanes Algorithm
# This algorithm runs in a single loop and keeps a running sum.
# if running sum ever drops below zero, we reset it to zero.
# its an optimal approach to find the maximum subarray sum.
## T.C -> O(n)
## S.C -> O(1)

def better_approach(nums):
    length = len(nums)
    current_sum = nums[0]
    running_sum = 0

    for i in range(length):
        running_sum += nums[i]
        current_sum = max(running_sum, current_sum)
        if running_sum < 0:
            running_sum = 0
    return current_sum



nums = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
print(brute_force(nums))
print(better_approach(nums))

File: cpp/miscellaneous_practice/LeetCode_53_Maximum_subarray.cpp

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// problem statement
//  we have to find the subarray wiith the maximum sum
//  sub array= array inside the array
//  [-2,1,-3,4,-1,2,1,-5,4]
//  [-2,1,-3] or [-5,4] or the array itself can be called as an subarray
// the sum should be talen in a continuos manner

// lets try an brute force approach using two loops
//  using max function to store the largest sum
//  we only have to return the largest subarray sum and not that subarray
// T.C -> O(n^2)
// S.C -> O(1)
int brute_force(vector<int> &nums)
{
    int len = nums.size();
    int maximum_sum = nums[0];

    for (int i = 0; i < len; i++)
    {
        int sum =0;
        for (int j = i; j < len; j++)
        {
            sum = sum + nums[j];
            maximum_sum = max(sum, maximum_sum);
        }
    }
    return maximum_sum;
}
 

// kadanes algo
// the algo is simple it simply runs in one loop and has a variable called current sum 
// the sum is set to zero whenever the its lower than zero

int better_approch(vector <int>& nums){
    int len=nums.size();
    int current_sum=nums[0];
    int sum=0;
    for(int i=0;i<len;i++){
        sum=sum+nums[i];
        current_sum=max(sum,current_sum);
        if(sum<0){
            sum=0;
        }
    }
    
    return current_sum;
}


int main()
{   vector <int> nums={-2,1,-3,4,-1,2,1,-5,4};
    cout<<brute_force(nums)<<endl;
    cout<<better_approch(nums)<<endl;
    return 0;
}